pilopt/src/lib.rs

```diff
@@ -686,6 +886,67 @@ fn constrained_to_constant<T: FieldElement>(
     None
 }
 
+/// Identifies witness columns that are constrained to a (multi)linear expression.
+fn constrained_to_linear<T: FieldElement>(
+    identity: &AlgebraicExpression<T>,
+    intermediate_columns: &BTreeMap<AlgebraicReferenceThin, AlgebraicExpression<T>>,
+) -> Option<((String, PolyID), AlgebraicExpression<T>)> {
+    // We require the constraint to be of the form `left = right`, represented as `left - right`
+    let (left, right) = if let AlgebraicExpression::BinaryOperation(AlgebraicBinaryOperation {
+        left,
+        op: AlgebraicBinaryOperator::Sub,
+        right,
+    }) = identity
+    {
+        (left, right)
+    } else {
+        return None;
+    };
+
+    // We require `left` to be a single, non-shifted, witness column `w`
```

Comment:
Should we also try with left and right swapped?

Reply:
done